### ISSUE TITLE
compiling error on firelfy3399 ubuntu platform

### ISSUE THREAD

**Issue by kaishijeng** [sat8/YoloOCLInference#1]
I got the following error when I do make after cmake:

/usr/bin/ld: CMakeFiles/YoloOCLInference.dir/src/YoloOCLMain.cpp.o: undefined reference to symbol '_ZN2cv6String10deallocateEv'
//usr/local/lib/libopencv_core.so.3.3: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
CMakeFiles/YoloOCLInference.dir/build.make:381: recipe for target 'YoloOCLInference' failed
make[2]: *** [YoloOCLInference] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/YoloOCLInference.dir/all' failed
make[1]: *** [CMakeFiles/YoloOCLInference.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

Any idea which link library is missing in your CMakeLists.txt

**Comment by sat8** [comment]
@kaishijeng I have been able to compile & run on Asus Tinkerboard running Yocto so as long as OpenCV installation is good, the project should compile. If you look in CMakelists.txt, then **opencv_core opencv_highgui opencv_imgproc opencv_imgcodecs opencv_videoio** are what the linker expects for Linux build. I also see you may have OpenCV 3.3.0 installed. Perhaps you have more than one version of  OpenCV on this system? https://forum.qt.io/topic/36421/error-time-usr-bin-ld-warning-libopencv_core-so-3-0 (the author says he had to reinstall OpenCV, may be worth a try...)

**Comment by kaishijeng** [comment]
You are right. After deleting opencv2.4.9 from the system, I can compile it without an error

Thanks,

**Comment by kaishijeng** [comment]
opencl 1.2 in my system seems not supporting unroll when I run. Can this be disabled?

 ./YoloOCLInference -input pedestrians.jpg -save 1                        

YoloOCLInference Started..
CL_COMPUTE DEVICES: 1
CL_DEVICE_ID: 0x7f854dadd8
CL_DEVICE_NAME:: Mali-T860
CL_DEVICE_VENDOR:: ARM
CL_DRIVER_VERSION:: 1.2
CL_DEVICE_VERSION:: OpenCL 1.2 v1.r13p0-00rel0-git(a4271c9).31ba04af2d3c01618138bef3aed66c2c
CL_DEVICE_OPENCL_C_VERSION:: OpenCL C 1.2 v1.r13p0-00rel0-git(a4271c9).31ba04af2d3c01618138bef3aed66c2c
CL_DEVICE_TYPE::CL_DEVICE_TYPE_GPU
CL_DEVICE_MAX_COMPUTE_UNITS: 4
clCreateProgramWithSource success

Build Log :<source>:16:26: error: OpenCL extension 'cl_nv_pragma_unroll' is not supported
#pragma OPENCL EXTENSION cl_nv_pragma_unroll : enable
                         ^

<source>:156:33: warning: unsequenced modification and access to 'offset'
                output[filter * size + (offset++)] = output[filter * size + offset] * scales[filter] + biases[filter];
                                              ^                             ~~~~~~

<source>:252:27: warning: array index 10 is past the end of the array (which contains 10 elements)
        maxValue = max(maxValue, localInput[10]);
                                 ^          ~~

<source>:228:2: note: array 'localInput' declared here
        float8 localInput[10];
        ^

<source>:293:27: warning: array index 10 is past the end of the array (which contains 10 elements)
        maxValue = max(maxValue, localInput[10]);
                                 ^          ~~

<source>:269:2: note: array 'localInput' declared here
        float8 localInput[10];
        ^

error: Compiler frontend failed (error code 59)

clGetProgramBuildInfo() ERROR - CL_BUILD_PROGRAM_FAILURE
Number of kernel Arguments : -1 image2columarray3x3
Number of kernel Arguments : 0 image2columarray1x1
Number of kernel Arguments : 0 normalizearray
Number of kernel Arguments : 0 scalebias
Number of kernel Arguments : 0 addbias
Number of kernel Arguments : 0 scaleaddbias
Number of kernel Arguments : 0 normscaleaddbias
Number of kernel Arguments : 0 leakyactivatearray
Number of kernel Arguments : 0 linearactivatearray
Number of kernel Arguments : 0 flattenarray
Number of kernel Arguments : 0 softmax
Number of kernel Arguments : 0 maxpool
Number of kernel Arguments : 0 resetarray
    0 conv     16  3 x 3 / 1   416 x 416 x   3   ->   416 x 416 x  16
    1 max          2 x 2 / 2   416 x 416 x  16   ->   208 x 208 x  16
    2 conv     32  3 x 3 / 1   208 x 208 x  16   ->   208 x 208 x  32
    3 max          2 x 2 / 2   208 x 208 x  32   ->   104 x 104 x  32
    4 conv     64  3 x 3 / 1   104 x 104 x  32   ->   104 x 104 x  64
    5 max          2 x 2 / 2   104 x 104 x  64   ->    52 x  52 x  64
    6 conv    128  3 x 3 / 1    52 x  52 x  64   ->    52 x  52 x 128
    7 max          2 x 2 / 2    52 x  52 x 128   ->    26 x  26 x 128
    8 conv    256  3 x 3 / 1    26 x  26 x 128   ->    26 x  26 x 256
    9 max          2 x 2 / 2    26 x  26 x 256   ->    13 x  13 x 256
   10 conv    512  3 x 3 / 1    13 x  13 x 256   ->    13 x  13 x 512
   11 max          2 x 2 / 1    13 x  13 x 512   ->    13 x  13 x 512
   12 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   13 conv   1024  3 x 3 / 1    13 x  13 x1024   ->    13 x  13 x1024
   14 conv    425  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x 425
   15 detection
Error creating directory output!
Error: argument index out of range

**Comment by sat8** [comment]
Yes, I noticed this too. This needs to be fixed properly but for now, you can comment out the line **#pragma OPENCL EXTENSION cl_nv_pragma_unroll : enable** in src/Kernels/DeepNNFP32.cl file. That should fix it.

**Comment by kaishijeng** [comment]
It works. However, FPS is quite low and it is slower, 0.77 FPS
 than Neon-optimized Tensorflow or Caffe on firefly3399. I will expect
opencl on GPU should be faster than Neon acceleration. Is there a way to
speed up?

Thanks,

**Comment by sat8** [comment]
gr8. Yes, the performance on ARM is not that great compared to desktop GPU cards. Presently I am spending time looking into it and I need to find the choke points that is causing a performance drop & optimise them. The present code is quite well optimised for desktop cards as it attains good processing speeds compared to CUDA version of tiny-yolo. I also have plans to use FP16 & INT8 versions to gain some processing speed. Stay tuned..

**Comment by sat8** [comment]
@kaishijeng I may have found the issue. Based on profiling OpenCL kernels, it seems GEMM has become a choke point on ARM boards. I have raised an [issue](https://github.com/CNugteren/CLBlast/issues/205) with the author of CLBLast library. I will try [ARM compute library](https://github.com/ARM-software/ComputeLibrary) if I find some free time next week.

**Comment by kaishijeng** [comment]
Will wait for your guidance.
By the way, CLBlast has a tuning feature which can use to tune its library
for each platform
Will this help?
.

Thanks

**Comment by sat8** [comment]
Yes, I have come to know about the tuning feature (On an AMD RX580 I did not get good results compared to stock settings in CLBlast database). Having said that, the author of CLBlast [hints](https://github.com/CNugteren/CLBlast/issues/128)  that the issue is due to the inefficiency of ARM OpenCL compiler compared to other OpenCL compilers. This is a grey area for me and fortunately few folks have offered suggestions/help. Watch this space for more info in coming days..

**Comment by kaishijeng** [comment]
I haven't seen good results of OpenCL on firefly 3399 on
ComputeLibrary:


https://github.com/OAID/MXNetOnACL/blob/master/acl_openailab/performance_report.pdf
)

and CLBlast:


https://github.com/dividiti/ck-caffe-firefly-rk3399/blob/master/script/batch_size-libs-models/analysis.20170531.ipynb

.
They are always worse than Neon-based acceleration. It is quite
disappointed.

Thanks,

**Comment by sat8** [comment]
@kaishijeng thnx 4 this. I had some advice on tuning CLBlast but didn’t really find noticeable improvements on ARM. So I need to do 2 things going forward. One is to make sure to be conprehensiv on tunjng process. The second, which may be right way going forward is to integrate  ACL neon for ARM compilation. I hope to get ACL neon going by end of next week(tentative).
